java/vector/src/main/java/org/apache/arrow/vector/complex/impl/PromotableWriter.java

```diff
@@ -268,32 +270,18 @@ private FieldWriter promoteToUnion() {
 
   @Override
   public void write(DecimalHolder holder) {
-    // Infer decimal scale and precision
-    if (arrowType == null) {
-      arrowType = new ArrowType.Decimal(MAX_DECIMAL_PRECISION, holder.scale);
-    }
-
-    getWriter(MinorType.DECIMAL, arrowType).write(holder);
+    getWriter(MinorType.DECIMAL, new ArrowType.Decimal(MAX_DECIMAL_PRECISION, holder.scale)).write(holder);
```

Comment:
If decimalVector is not already added, it will get added during the call to getWriter() which will create the vector based on the precision and scale of this arrowType. If the precision passed is holder.precision, any subsequent writes to this vector with precision greater than this precision will throw an error (eg, after adding writeDecimal(new BigDecimal(5)), writeDecimal(new BigDecimal(15)) will throw error due to precision mismatch). So its best to pass MAX_DECIMAL_PRECISION for such cases.